ocamltest

shindere opened the pull request:
This PR introduces the ocamltest tool which aims at making it easier to
compile, execute and check test results of the OCaml testsuite, in order
to finally replace the makefiles that are currently in use.

A bit of documentation is available in ocamltest/README.

This branch, including the README, is very much work in progress.

Any contribution is very welcome: PRs against this branch to help migrate
tests, comments about the ocamltest tool...

alainfrisch commented:
Did you consider using attributes instead of a custom language lexer/parser for TEST blocks?

shindere commented:
Alain Frisch (2016/07/11 07:59 -0700):

> Did you consider using attributes instead of a custom language
> lexer/parser for TEST blocks?

Well, not really so far, to be honest.

I can see two questions:
1. Are attribute trustworthy enough compared to lex/yacc? I mean,
   lex/yacc have been used to compile the compiler, so they are know to
   work, or at least it is likely that they work correctly. Can the same be
   said about attributes?

Second, I have to say I am not familiar with attributes. Thus, I don't
know whether they would provide the same level of freedom in the form of
the TEST block. At the moment the block can contain tst trees, with the
possibility for each test to define its own environment variables. I
don't know whtehr that would be doable with attributes, at which cost
for the developer and for users. And I don't know either what the mpact
would be on ocamltest's overall architecture.

Octachron commented:
After reading the readme, I am not sure which tests can be converted to the ocamltest format.
Am I wrong to think that none of the tool-related (e.g. testsuite/tests/tool-ocamldoc-open ) tests, nor the tests based on expect_test (e.g. testsuite/tests/typing-gadts) , nor any test based on Makefile.okbad (e.g. testsuite/tests/typing-recmod), nor any specific Makefiles can be converted as for now?

chambart commented on .merlin:
```diff
@@ -1,56 +0,0 @@
-S ./asmcomp
```
Why remove the .merlin file ?

shindere replied:
That was a mistake (thanks for having spotted it!) which has been fixed
by editing the commit. So the branch needs to be checked out again.

shindere commented:
Indeed, the tests you mention cannot be migrated right now. ocamltest
needs to be extended to support msuch tests. I'lyl work on that self but for
that, too, help is welcome.

Octachron commented:
If I understand correctly all specific makefiles would need to be rewritten within ocamltest (i.e. a new compiler-specific tool), and all tests would need to be given a specification in TSL, a new compiler-specific DSL. 

This seems to add quite a bit of overhead compared to the existing makefile-based situation. Could you comment in more detail what would be the advantages of such transition?

shindere commented:
Florian Angeletti (2016/07/11 14:31 -0700):

> If I understand correctly all specific makefiles would need to be
> rewritten within ocamltest (i.e. a new compiler-specific tool),

You are correct, except for the "compiler-specific" aspect. The tool has
indeed been written with the idea that it could be re-used by other
projects with similar needs, especially in the OCaml ecosystem.

Not sure this reusability aspect has been reached yet, but it is at
least a target.

> and all tests would need to be given a specification in TSL, a new
> compiler-specific DSL.

That's correct, too.

> This seems to add quite a bit of overhead compared to the existing
> makefile-based situation.

I think there are at least a few developers who may have a different
opinion on this. It indeed seems that the current makefile-based
approach is not too easy to maintain or to extend.

> Could you comment in more detail what would be the advantages of such
> transition?

I'd say having all the logic in one place, having it written in OCaml
rather than Makefile, having an easier to debug system. For instance
with the current system it is rather hard to figure out what goes wrong
when a test fails: the commands used to produce it are not logged, the
final binary, "program", will in general be overwritten by the next test
(except for the last test in a directory). And even getting to such
a result requires some hacking when adding tests, especially when
the tests are in a new directory.

dra27 commented:
What's the plan for Windows? The AppVeyor pass is a fake: see lines 5366-5441 of the log.

shindere commented:
@dra27 Many thanks for your interest in this PR, David.

I am very interested in making this work under Windows and welcome any
feedback especially on this topic, given that I have no experience at
all with Windows programing.

Where is the log you mention visible, please?

xavierleroy commented on ocamltest/run.c:
```diff
@@ -0,0 +1,301 @@
+        realpath_error(settings->stdout_filename);
+      if ((realpath(settings->stderr_filename, stderr_realpath) == NULL) && (errno != ENOENT))
+        realpath_error(settings->stderr_filename);
+#endif /* __GLIBC__ */
```
It took me a while to understand what this code is doing.  I suggest moving it away in a function

```
int paths_same_file(const char * path1, const char * path2)
```

xavierleroy replied:
An alternate approach that doesn't use `realpath`:
- open both stdout_filename and stderr_filename
- `fstat` the  two resulting file descriptors
- if they have identical `st_dev` and `st_ino`, close the second fd and make it equal to the first fd

shindere replied:
Xavier Leroy (2016/07/12 00:43 -0700):

> An alternate approach that doesn't use `realpath`:
> - open both stdout_filename and stderr_filename
> - `fstat` the  two resulting file descriptors
> - if they have identical `st_dev` and `st_ino`, close the second fd
> - and make it equal to the first fd

Thanks a lot for this suggestion! It looks much simpler than the current
implementation and thus rather appealing.

Just a question before implementing this:

Isn't there going to be a loss in portability?

I have no idea how the st_ino is defined on a non-FAT filesystem...

mrvn replied:
The `st_dev` check garantees you are comparing inodes on the same filesystem. The inode is unique within a filesystem, any filesystem. It's how the VFS (virtual file system) interface in any unix like system works.

shindere replied:
Goswin von Brederlow (2016/07/12 02:51 -0700):

> The `st_dev` check garantees you are comparing inodes on the same
> filesystem. The inode is unique within a filesystem, any filesystem.
> It's how the VFS (virtual file system) interface in any unix like
> system works.

I meant: how does all this work on Windows, on a FAT or NTFS filesystem?

xavierleroy replied:
Under Windows, with the Win32 API, you would use GetFileInformationByHandle instead of fstat.  (Note that run.c needs two completely different implementations anyway, one for Unix and the other for Win32.)

Under Cygwin, I hope the fstat emulation is good enough to produce pseudo st_ino numbers that are unique within a file system.  (To be checked.)

Under Linux, if you mount a FAT or NTFS file system, I'm not sure st_ino numbers are meaningful, but do we really care for this use case?

dra27 replied:
Indeed - you cannot rely on st_dev and st_ino on the Microsoft implementation of stat. Alternatively, on Windows, you could use my reimplementation of C stat (see `otherlibs/win32unix/stat.c`) which does (well, certainly _should_) give meaningful st_ino and st_dev numbers.

shindere replied:
David Allsopp (2016/07/13 09:47 -0700):

> Indeed - you cannot rely on st_dev and st_ino on the Microsoft
> implementation of stat. Alternatively, on Windows, you could use my
> reimplementation of C stat (see `otherlibs/win32unix/stat.c`) which
> does (well, certainly _should_) give meaningful st_ino and st_dev
> numbers.

So am I correct that the current implementation is actually more
portable?

Wouldn't it be better, then, to keep it, it its own function as
@xavierleroy suggested?

shindere replied:
Xavier Leroy (2016/07/12 00:39 -0700):

> > -      stdout_realpath = realpath(settings->stdout_filename, NULL);
> > -      if (stdout_realpath == NULL)
> > -        realpath_error(settings->stdout_filename);
> > -      stderr_realpath = realpath(settings->stderr_filename, NULL);
> > -      if ( (stderr_realpath == NULL)  && (errno != ENOENT) )
> > -      {
> > -        free(stdout_realpath);
> > -        realpath_error(settings->stderr_filename);
> > -      }
> >   +#else
> > -      char stdout_realpath[PATH_MAX], stderr_realpath[PATH_MAX];
> > -      if (realpath(settings->stdout_filename, stdout_realpath) == NULL)
> > -        realpath_error(settings->stdout_filename);
> > -      if ((realpath(settings->stderr_filename, stderr_realpath) == NULL) && (errno != ENOENT))
> > -        realpath_error(settings->stderr_filename);
> >   +#endif /\* **GLIBC** */
> 
> It took me a while to understand what this code is doing.  I suggest moving it away in a function
> 
> ```
> int paths_same_file(const char * path1, const char * path2)
> ```

Thanks. This has finally been done.
It turned out that the settings also needed to be propagated.

mrvn commented:
Why is this written in C when we have a perfectly fine functional language with yacc and lex to use?

shindere commented:
Goswin von Brederlow (2016/07/12 01:24 -0700):

> Why is this written in C when we have a perfectly fine functional
> language with yacc and lex to use?

There is only the part that runs programs that has been written in C. It
supports redirections, timeouts and things like that. As
ocamltest/README explains, it is not possible to use the Unix library at
this stage because it has not been tested yet.

alainfrisch commented:
I understand the desire to reduce the "trust base", but well, Unix has been tested for years and if some bad regression is introduced, it is likely to break tests, not make them succeed silently, so I don't see it as a real issue to rely on Unix in the testsuite.  In a sense, the new code in the testsuite (C or not) will not be more tested than Unix, so there is no reason to expect it to be more trustworthy than Unix.

Making the testsuite more portable and easier to build seems more important at this stage.

shindere commented:
Alain Frisch (2016/07/12 01:37 -0700):

> I understand the desire to reduce the "trust base", but well, Unix has
> been tested for years and if some bad regression is introduced, it is
> likely to break tests, not make them succeed silently, so I don't see
> it as a real issue to rely on Unix in the testsuite.  In a sense, the
> new code in the testsuite (C or not) will not be more tested than
> Unix, so there is no reason to expect it to be more trustworthy than
> Unix.

I definitely agree with this. And while I was writing this code it felt
odd, I couldn't convince myself that this way was the right one.

But would it really be possible / simpler to achieve what run is
doing in pure OCaml?

alainfrisch commented:
> But would it really be possible / simpler to achieve what run is doing in pure OCaml?

I did not check, but I do not see upfront why a "testsuite driver" would depend on features not available in Unix (and even in the subset available under Windows).  Basically, you need to launch processes and monitor (with Unix.select) their stderr/stdout and get their exit code; anything else?

xavierleroy commented:
I disagree that the Unix interface is well tested, esp. its Win32 implementation.  (While working on PR#650 I noticed a number of places where the behavior is not obvious to me and I don't know if anyone exercised them ever.)  Plus, the "launch external process with a timeout" primitive that ocamltest needs would be nearly impossible to implement under Win32 using only what is available in the OCaml Unix interface.  So, I think this primitive must be implemented in C anyway.

alainfrisch commented:
>  Plus, the "launch external process with a timeout" primitive that ocamltest needs would be nearly impossible to implement under Win32 using only what is available in the OCaml Unix interface.

Can you elaborate?  `Unix.kill` works under Win32 for killing a process.  The timeout would be managed by the driver (that would rely on `select` I guess).

xavierleroy commented:
@alainfrisch: What is needed here is roughly `waitpid` with a timeout.  `select` won't help as it selects on file descriptors, not process ID.  The Win32 API accessed directly from C provides the desired functionality with a single system call (`WaitForSingleObject`).  Restricting ourselves to the Unix API of OCaml, the only solutions I see involve signals (SIGALRM interrupting `waitpid` or SIGCHLD interrupting `sleep`) or `fork`, none of which is implemented under Win32.  The best I could do is "semi-active polling": `waitpid` with `WNOHANG`, then sleep for a short while, and repeat.  Not very satisfying.

At any rate, I think this question "in which language is the process launcher implemented" is inessential.  I'm more concerned about what needs to be done to support the full variety of tests that we have today or will need shortly.

c-cube commented:
I think this kind of primitive should still be used from OCaml, possibly after being added to `Unix`. Currently I cannot really use some high-level subprocess-related functions from `Unix` because I need the PID, a timeout and the return code along with stdout/err (no way to get all this).

shindere commented:
Branch rebased on latest trunk (1ae28b98cf).
ocamltest extended to support expect tests.
expect tests migrated.

There is just one test that has a strange behaviour:
testsuite/tests/typing-modules/printing.ml
Giving it to expect_test displays this line:

module M : sig module N : sig ... end end

(the line appears both with and without -principal)

The test passes, in the sense that printing.ml.corrected.corrected is
identical to printing.ml, but I'm wondering whether it is expected that
this line is displayed.

Any hint @diml perhaps?

ghost commented:
`expect_test` wasn't capturing `Format.{std,err}_formatter`, so the output of `#`-directives was going to the terminal. I changed this in 703d786d17025c6126092de039b140c4b5255e4d

mrvn commented:
Under unix a simple fix would be to keep a pipe to the process open and select on it. When the process dies the pipe gets closed and select returns. But windows has no pipes, right?

shindere commented:
Thanks alot @diml for your prompt and helpful response!

dra27 commented:
@mvrn: Windows very much has pipes! But passing the handles to the child processes is much clunkier than on Unix. It would probably be easier to co-opt selecting on one of the standard handles instead, if that can be made to work?

dra27 commented:
@shindere: it's the AppVeyor CI log from the "All checks have passed" section. The initial fix is probably simple - it's just that `Makefile.nt` doesn't build ocamltest. I'd love to be offering more pragmatic support at this stage, but there're only so many hours in the day - but I would think you will have a much better time getting Windows working at the "60 tests are ported stage" and going from there than waiting until 500 tests are ported and then trying to do it!
For an easier start, just using one of the mingw ports keeps the build-system gcc-like but forces you not to rely on Posix/libc features. Getting the msvc ports to work is usually then just a matter of adapting command/filename invocations.

gerdstolpmann commented:
This is interesting work. I'm wondering whether we could add one badly needed feature: run the testsuite for cross-compile targets. Some months ago I spent some hours on improving the old testsuite so that you could do the following:
- compile the tests on the host system for the target system
- the tests on the target system run in server mode: a super-simple protocol accepts TCP connections (communicating which tests to run, and stdin/stdout)
- the tests are linked together into a single executable (that's not a "must-have" but it eases running the tests)
- the test driver on the host system selects the tests, and invokes the remote test server, and checks the result

You may wonder why the client/server architecture was preferred. The point is that you cannot officially log in on many ARM devices (without jailbreaks which are not available for all OS versions). With the client/server approach you can use the official methods of running development executables on such devices. I was doing that for iOS, and added a simple xcode wrapper that calls from Objective-C into OCaml so that the test server looked like a normal app.

Not every test may be suited for this method, but it would be incredibly helpful for testing on such devices if we could select a bunch of tests and run them in client/server mode, as outlined.

shindere commented:
Many thanks for all your feedback, @dra27

David Allsopp (2016/07/13 10:38 -0700):

> @shindere: it's the AppVeyor CI log from the "All checks have passed"
> section.

Okay thanks, I see now.

> The initial fix is probably simple - it's just that `Makefile.nt`
> doesn't build ocamltest.

This should have been fixed by now.

> I'd love to be offering more pragmatic support at this stage, but
> there're only so many hours in the day

That's fine, your help is already very precious and very appreciated.

> but I would think you will have a much better time getting Windows
> working at the "60 tests are ported stage" and going from there than
> waiting until 500 tests are ported and then trying to do it!

Fully agreed. About 130 tests have been ported so far so it seems
definitely a good time to think about Windows seriously.

> For an easier start, just using one of the mingw ports keeps the
> build-system gcc-like but forces you not to rely on Posix/libc
> features.

Okay.

> Getting the msvc ports to work is usually then just a matter of
> adapting command/filename invocations.

Regarding this: so far I have not been able to find a way to
access the extension of executables from within an OCaml program.
Am I correct that this is not available yet?
Would it be a good idea to add an exe_extension value to config.ml?

I'm williing to submit a PR if that is considered a good idea.

dra27 commented:
There's no function for adding the executable extension, but you can rely on `Sys.win32 = true` as a test for adding `.exe` (ignoring Cygwin here - mostly, Cygwin magically handles the `.exe` for you).

Improving that properly (e.g. `Filename.add_executable_extension` or something) will require consideration for cross-compiling, which is likely to slow you down! Same goes for improving the value in `config.ml` - although you can sort of get it by analysing `default_executable_name` (though doing this has been criticised before (sorry - the inadequacy of the caml-list archives means I can't find the thread!)

shindere commented:
Thanks a lot for all your feedback, which I find
definitely interesting, @gerdstolpmann.

I can't promise to take it into account immediately, because I find
there is already a lot of work to do to make the tests work and to
extend ocamltest, but I think in the long run your suggestion should
definitely be implemented, one way or another.

Moreover, improving cross-compilation is part of the work I am supposed
to do with Gallium so I really hope I can come back to your suggestion
when I focus more specifically on cross-compilation issues.

shindere commented:
Jérémie Dimino (2016/07/12 06:57 -0700):

> `expect_test` wasn't capturing `Format.{std,err}_formatter`, so the
> output of `#`-directives was going to the terminal. I changed this in
> 703d786d17025c6126092de039b140c4b5255e4d

Many thanks for your prompt response and the fix!

damiendoligez commented on ocamltest/README:
```diff
@@ -0,0 +1,181 @@
+Define the following environment variables:
+
+export OCAMLSRCDIR=/path/to/your/clone/of/ocaml/sources
+export OCAMLTESTDIR=/tmp/ocamltest
```
These two variables should be set by default by the configure/makefile infrastructure.

damiendoligez commented on ocamltest/builtin_actions.ml:
```diff
@@ -0,0 +1,880 @@
+  Printf.sprintf "%s: command\n%s\nfailed with exit code %d"
+    what commandline exitcode
+
+let make_file_name name ext = String.concat "." [name; ext]
```
Out of curiosity: why do you use `String.concat` here, `sprintf` in the previous function, and `^` in `backends.ml` ?

damiendoligez commented on ocamltest/run_unix.c:
```diff
@@ -0,0 +1,273 @@
+error_with_location(__FILE__, __LINE__, settings, msg, ## __VA_ARGS__)
+
+/*
+  Note: the ## __VA_ARGS__ construct is gcc specific.
```
I presume it also works in clang, but people with more exotic compilers (icc, Sun) will be sad.

damiendoligez commented on ocamltest/selftests/bug.c:
```diff
@@ -0,0 +1,17 @@
+{
+  char *p = NULL;
+  if ( chdir("/tmp") < 0) perror("chdir");
+  *p = 'a';
```
This is not guaranteed to dump a core... You should use the `abort` function from the C standard library.

damiendoligez commented on ocamltest/selftests/bug2.c:
```diff
@@ -0,0 +1,16 @@
+{
+  char *p = NULL;
+  if ( chdir("/") < 0) perror("chdir");
+  *p = 'a';
```
Same remark about `abort`.

damiendoligez reviewed (commented):
A few comments. I will also submit a pull request for typos and grammar and talk to @shindere in person.

shindere committed:
ocamltest: the test driver for the OCaml compiler

This commit contains:

- The initial version of the tool itself, in the ocamltest directory
- The required additions to the main .gitignore and .merlin files.

The integration of ocamltest in OCaml's main build system and its use
in the testsuite are not part of this commit.

Credits go to David Allsopp for
- Extending the computation of ocamlsrcdir to the Windows case
- Making ocamltest compile when flexlink
is being bootstrapped.

shindere committed:
Integrate the build of ocamltest to the main build system

This commit ensures ocamltest.byte and ocamltest.opt get built along with
the other tools.

shindere committed:
configure should define the CPP build variable

The CPP variable tells the build system how to invoke the C preprocessor.

Before this commit, this variable was defined by the Windows Makefiles
but not by the Makefile generated by configure. This commit fixes this.

This commit also fixes a few makefiles: to compute dependencies for C files,
they incorrectly called $(CPP) -MM instead of $(CC) -MM.

shindere committed:
Make testsuite able to run ocamltest-based tests

This commit enhances the test harness so that it becomes able to
run ocamltest-based tests besides to the legacy tests it already runs.

This is achieved as follows:

Modification of targets in testsuite/Makefile:

- make legacy runs the legacy tests

- make new runs the ocamltest-based tests

  This uses ocamltest.opt if present and falls back to ocamltest
  (bytecode version) otherwise.

- make all runs both the legacy tests and the ocamltest-based tests

- make clean also removes the _ocamltest directories created by
  ocamltest to store test outputs

To know which tests to run, the testsuite/tests directory tree is
explored recursively as follows. For each of its subdirectory d:

- If d contains a Makefile, then d is assumed to contain legacy tests

- If d contains an ocamltests file, then d is assumed to contain
  ocamltest-based tests whose names are listed in the ocamltests file,
  one per line.

- If none of these files is found, then the subdirectories of d are explored
  recursively.

Thanks to David Allsopp for having provided a patch to let ocamltest
be run when flexlink is being bootstrapped.

shindere committed:
testsuite: provide the testing component as a library

A few tests rely on a Testing module located in testsuite/lib.

The legacy testing infrastructure links it into the programs that
use it as a module, but in the context of ocamltest it seems better
to provide testing as a library.

This commit thus modifies the build system of the testsuite to
also build testing as a library, both in bytecode and native formats.

shindere committed:
Migrate testsuite/tests/basic-io

shindere committed:
Testsuite infrastructure: add .PHONY declarations where appropriate

shindere committed:
Migrate testsuite/tests/array-functions

shindere committed:
Migrate testsuite/tests/basic-io-2

shindere committed:
Migrate testsuite/tests/basic-modules

shindere committed:
Migrate testsuite/tests/basic-multdef

shindere committed:
Migrate testsuite/tests/basic-private

shindere commented:
This PR has been rebased on trunk.

It contains less migrated than a year ago because ocamltest has evolved
a bit meanwhile and I was in lack of time to make sure that all the
tests that did work in the past do still work.

They will hence be migrated in subsequent pull requests.

damiendoligez commented on ocamltest/filetype.ml:
```diff
@@ -0,0 +1,69 @@
+  | Implementation
+  | Interface
+  | C
+  | C_minor
```
minor nit: this should be C_minusminus rather than C_minor

damiendoligez commented on ocamltest/testlib.ml:
```diff
@@ -0,0 +1,134 @@
+  begin
+    close_in chan;
+    failwith
+      ("The fle " ^ filename ^ " is too large t be loaded into a string")
```
Two typos in this sentence: "fle" should be "file" and "t" should be "to".

shindere commented:
@damiendoligez both fixed, thanks.

shindere commented:
Does someone (perhaps you, @dra27 ?) understand why AppVeyor fails?

It tries to compile the bytecode version of ocamltest with this command:

../byterun/ocamlrun ../ocamlc -g -nostdlib -I ../utils -I ../parsing -I ../stdlib -I ../compilerlibs -strict-sequence -safe-string -strict-formats -w +a-4-9-41-42-44-45-48 -warn-error A -custom ocamlcommon.cma -o ocamltest.exe run_win32.o run_stubs.o ocamltest_config.cmo testlib.cmo run_command.cmo filetype.cmo filecompare.cmo backends.cmo variables.cmo environments.cmo builtin_variables.cmo builtin_modifiers.cmo actions.cmo builtin_actions.cmo tests.cmo builtin_tests.cmo tsl_ast.cmo tsl_parser.cmo tsl_lexer.cmo tsl_semantics.cmo options.cmo main.cmo

And this produces the following error message:

** Fatal error: Cannot find file "flexdll_mingw.o"

mshinwell commented:
I've asked @dra27 to chime in on this, and he will do so shortly.

shindere commented:
Mark Shinwell (2017/09/15 09:22 +0000):
> I've asked @dra27 to chime in on this, and he will do so shortly.

Thanks a lot, Mark!

dra27 commented:
You're missing the logic to set `OCAML_FLEXLINK` (cf. https://github.com/ocaml/ocaml/blob/trunk/Makefile#L301). I think you need to add this to your definition for ocamlc and ocamlopt: `OCAML_FLEXLINK="boot/ocamlrun flexdll/flexlink.exe"`

mingw is tested using a bootstrapped flexlink - since that variable is not set, you're picking up the system installed flexlink but that (intentionally) does not have the mingw objects.

shindere commented:
Okay thanks a lot, @dra27 !
Will look into this immediately.

Do you understand why ocamldoc compiles, though?

It does not seem more flexdll-aware than ocamltest and still, it
works...

dra27 commented:
It is aware - https://github.com/ocaml/ocaml/blob/trunk/ocamldoc/Makefile#L27-L32

dra27 commented:
I'm just testing a patch, btw!

shindere commented:
David Allsopp (2017/09/15 09:40 +0000):
> It is aware -
> https://github.com/ocaml/ocaml/blob/trunk/ocamldoc/Makefile#L27-L32

Ah indeed, but see, in presence of the flexdll submodule, only the
definition of OCAMLOPT is affected, right? Can you understand how things
work for ocamlc then? Because for ocamltset, it is the compilation to
bytecode which is broken!

>

dra27 commented:
You build with `-custom` so require flexlink - compilation without `-custom` doesn't need it (I think...)

shindere commented:
David Allsopp (2017/09/15 09:43 +0000):
> I'm just testing a patch, btw!

Okay thanks!

shindere commented:
David Allsopp (2017/09/15 02:45 -0700):
> You build with `-custom` so require flexlink - compilation without
> `-custom` doesn't need it (I think...)

Ah, and -custom is mandatory because there are C bits, right?

dra27 commented:
You'd have to dynamically link the C stuff otherwise, yes - so better to use `-custom` which works everywhere

dra27 commented:
Do you mind if I push a commit directly to the branch, or shall I post the patch? (I'm still waiting for world.opt to finish)

shindere commented:
David Allsopp (2017/09/15 02:50 -0700):
> Do you mind if I push a commit directly to the branch, or shall I post
> the patch? (I'm still waiting for world.opt to finish)

If you don't mind, I suggest the following:

- You push to the branch

- I rewrite the history to squash your commit in the one that introduces
ocamltset and

- I credit you in the commit message for having added the flexdll
support

Is that okay with you?

(I'm concerned about having the cleanest possible history)

dra27 commented:
That's absolutely fine - I'm more worried about it working than the credit! I can't push to your branch (perhaps the checkbox to allow maintainers to push is not checked) but here's the change:

```diff
diff --git a/ocamltest/Makefile b/ocamltest/Makefile
index 250b2388d..612d20c41 100644
--- a/ocamltest/Makefile
+++ b/ocamltest/Makefile
@@ -98,9 +98,19 @@ flags = -g -nostdlib $(include_directories) \
   -strict-sequence -safe-string -strict-formats \
   -w +a-4-9-41-42-44-45-48 -warn-error A

-ocamlc := ../byterun/ocamlrun ../ocamlc $(flags)
+ifeq "$(UNIX_OR_WIN32)" "unix"
+FLEXLINK_ENV=
+else # Windows
+  ifeq "$(wildcard ../flexdll/Makefile)" ""
+    FLEXLINK_ENV=
+  else
+    FLEXLINK_ENV=OCAML_FLEXLINK="../boot/ocamlrun ../flexdll/flexlink.exe"
+  endif
+endif
+
+ocamlc := $(FLEXLINK_ENV) ../byterun/ocamlrun ../ocamlc $(flags)

-ocamlopt := ../byterun/ocamlrun ../ocamlopt $(flags)
+ocamlopt := $(FLEXLINK_ENV) ../byterun/ocamlrun ../ocamlopt $(flags)
```

The testsuite is still running on my box, but that certainly builds for a bootstrapped mingw

dra27 commented:
Hmm - this is not working on my machine. I've had to do a temporary workaround to deal with `\r` problems (that can be fixed later - that's just to do with my using proper Windows git clones) but the only test which is passing is pr6322 if_not_safe_string. I'm trying to investigate further...

shindere commented:
David Allsopp (2017/09/15 10:34 +0000):
> Hmm - this is not working on my machine. I've had to do a temporary
> workaround to deal with `\r` problems (that can be fixed later -
> that's just to do with my using proper Windows git clones) but the
> only test which is passing is pr6322 if_not_safe_string. I'm trying to
> investigate further...

You mean the only text that *passes* of *fails* ?

dra27 commented:
I meant that only 1 test was passing - all the rest failed. I'm about to do a GPR against your branch with fixes... give me a minute!

shindere commented:
David Allsopp (2017/09/15 09:57 +0000):
> That's absolutely fine - I'm more worried about it working than the credit! I can't push to your branch (perhaps the checkbox to allow maintainers to push is not checked) but here's the change:

[...]

Thanks, aplied and pushed.

dra27 commented:
I'm still seeing failures, I'm afraid - investigating further!

mshinwell commented:
@shindere Is there anything else we can do to help towards getting this merged (on non-Windows platforms)?

shindere commented:
Mark Shinwell (2017/09/15 06:41 -0700):
> @shindere Is there anything else we can do to help towards getting
> this merged (on non-Windows platforms)?

I don't think so, thanks!

Will check with Damien how things go on the Mac side but apart from that
I hope we are close.

mshinwell commented:
I can help on the OS X side if you need extra hands too, just shout.

dra27 commented:
I've got 89 tests failing on Cygwin (that's likely to do with my CRLF checkout), 55 tests failing on MSVC (that needs looking at - I have a suspicion!) and the constprop test still failing on mingw which I'll continue to investigate...

shindere commented:
Mark Shinwell (2017/09/15 14:02 +0000):
> I can help on the OS X side if you need extra hands too, just shout.
Thanks. Actually the issue was that the Virtual machines had their DNS
servers misocnfigured.

This has now been fixed. I will fix one or two other things and re-run
Inria's CI everywhere afterwards.

Thanks again, Mark!

shindere commented:
constprop is weird, because it is preprocessed by the C preprocessor so
I am not so much surprised that it does not work. If we are unable to
make it work soon, I wouldn't mind keeping the basic directory where it
lives out of this PR, let's keep this possibility in mind.

@dra27 you can look at the test log and see whether the way the C
preprocessor is called is okay, I'm pretty sure that's where the error
comes from.

shindere commented:
Okay so I just started Inria's CI on this PR with all the
available slaves, that is: ocaml-arm-32/64, ocaml-cygwin-32/64,
ocaml-freebsd-64, ocaml-linux-32/64, ocaml-macos-10.9,
ocaml-macos-10.10, ocaml-mingw-32/64, ocaml-msvc-32/64,
ocaml-openbsd-32, ocaml-ppc-32/64, ocaml-zsystems (for those with CI
access it's job shindere-ocamltest build #5).

One thing: by default, ocamltest puts its outputs under /tmp/ocamltest.
This behaviour can be modified by defining the OCAMLTESTDIR environment
variable. I just added a ci-build modification to do that, otherwise
there would be collisions of those slaves where several jobs are
running.

Does this need to be taken into account on Travis / AppVeyor, or are we
sure that there will be one job per slave?

xavierleroy commented:
> by default, ocamltest puts its outputs under /tmp/ocamltest.

This is a problem not just with CI but also with multi-user machines (these still exist) or when a single user runs tests for several projects concurrently.  If you put things in `/tmp` they should have unique names, in the style of `Filename.temp_file`.

dra27 commented:
Ah - this could also be what I've seen on my Azure build slave, as it runs all 6 ports simultaneously.

shindere commented:
Xavier Leroy (2017/09/15 08:51 -0700):
> > by default, ocamltest puts its outputs under /tmp/ocamltest.
> 
> This is a problem not just with CI but also with multi-user machines
> (these still exist) or when a single user runs tests for several
> projects concurrently.  If you put things in `/tmp` they should have
> unique names, in the style of `Filename.temp_file`.

Ah that needs to be taken into account, indeed (but perhaps not before
the feature freeze...).

xavierleroy commented:
> by default, ocamltest puts its outputs under /tmp/ocamltest.

This is a problem not just with CI but also with multi-user machines (these still exist) or when a single user runs tests for several projects concurrently.  If you put things in `/tmp` they should have unique names, in the style of `Filename.temp_file`.

damiendoligez commented:
I've just suggested directly to @shindere that it should go into `_ocamltest` in the current directory, by analogy with build systems that put their stuff in `_build`. At worst, we can change that later.

shindere committed:
Tell git to ignore the _ocamltest directories created by ocamltest

shindere commented:
Xavier Leroy (2017/09/15 09:03 -0700):
> > by default, ocamltest puts its outputs under /tmp/ocamltest.
> 
> This is a problem not just with CI but also with multi-user machines
> (these still exist) or when a single user runs tests for several
> projects concurrently.  If you put things in `/tmp` they should have
> unique names, in the style of `Filename.temp_file`.

Following @damiendoligez 's suggestion, the output now goes to the
_ocamltest directory by default.

CI running again. I included a revert of
6fb2960d561fdec8c21aaf63d97e7c677e364c5e to be able to test without the
isatty patch.

shindere commented:
CI now passes for Unix builds.

On the two Cygwin ports, ocamltest fails to compile because of this error:

** Fatal error: Error while reading ../stdlib\libcamlrun.a: End_of_file
File "_none_", line 1:
Error: Error while building custom runtime system

Does anybody (@dra27?) know about this error?

I don't think it is FlexLink related.

Also, the way ocamlc is invoked is very similar to what is odne for
ocamldoc, which compiles.

FTR here is the command that produces the error message above:

../byterun/ocamlrun ../ocamlc -g -nostdlib -I ../utils -I ../parsing -I
../stdlib -I ../compilerlibs -strict-sequence -safe-string
-strict-formats -w +a-4-9-41-42-44-45-48 -warn-error A -custom
ocamlcommon.cma -o ocamltest.exe run_unix.o run_stubs.o
ocamltest_config.cmo testlib.cmo run_command.cmo filetype.cmo
filecompare.cmo backends.cmo variables.cmo environments.cmo
builtin_variables.cmo builtin_modifiers.cmo actions.cmo
builtin_actions.cmo tests.cmo builtin_tests.cmo tsl_ast.cmo
tsl_parser.cmo tsl_lexer.cmo tsl_semantics.cmo options.cmo main.cmo

Any hint appreciated.

nojb commented:
Did you try passing `-verbose` to figure out to figure out which external command fails ?

shindere commented:
Nicolás Ojeda Bär (2017/09/16 16:16 +0000):
> Did you try passing `-verbose` to figure out to figure out which
> external command fails ?

Nope, thanks!

shindere commented:
Btw since it's an End_of_file exception I don't think the failure is due
to an external command...

shindere commented:
Update: moving forward, slowly.

One of the reasons that caused trunk to fail on Inria's Windows slaves
was that flexlink.exe was not seen in the PATH. It had been added to the
environemnt, but in a file which is not read when Jenkins logs into the
slaves.

Running CI again to see what's still left.

dra27 commented:
@shindere - I'm very suspicious of the presence of a backslash in the path for libcamlrun.a for the Cygwin ports. I can't check the code here, but perhaps the Cygwin ports must use a Cygwin-compiled flexlink for -custom to work?

damiendoligez reviewed (approved):
LVGTM thanks!

gasche commented:
Currently, the AppVeyor CI machine seems broken (see for example [the fma build](https://ci.appveyor.com/project/avsm/ocaml/build/1.0.4243)), and I have the impression that this is ocamltest-related:

```
make[3]: Entering directory '/cygdrive/c/projects/ocaml/ocamltest'
cl -nologo -O2 -Gy- -MD -WX -D_CRT_SECURE_NO_DEPRECATE -DCAML_NAME_SPACE -DUNICODE -D_UNICODE -DWINDOWS_UNICODE=1 -I../byterun -DCAML_INTERNALS  -c run_win32.c
run_win32.c
run_win32.c(61): error C2220: warning treated as error - no 'object' file generated
run_win32.c(61): warning C4133: 'function': incompatible types - from 'const char *' to 'LPCWSTR'
run_win32.c(64): warning C4133: 'function': incompatible types - from 'char *' to 'LPWSTR'
run_win32.c(66): warning C4133: 'function': incompatible types - from 'char **' to 'LPWSTR *'
run_win32.c(88): warning C4133: 'function': incompatible types - from 'const char *' to 'LPCWSTR'
run_win32.c(91): warning C4133: 'function': incompatible types - from 'char *' to 'LPWSTR'
run_win32.c(93): warning C4133: 'function': incompatible types - from 'char **' to 'LPWSTR *'
run_win32.c(53): warning C4133: 'initializing': incompatible types - from 'char [5]' to 'LPCTSTR'
run_win32.c(140): warning C4133: 'function': incompatible types - from 'const char *' to 'LPCWSTR'
run_win32.c(157): warning C4133: 'function': incompatible types - from 'const char *' to 'LPCWSTR'
run_win32.c(250): warning C4133: 'function': incompatible types - from 'char *' to 'LPCWSTR'
run_win32.c(251): warning C4133: 'function': incompatible types - from 'char *' to 'LPWSTR'
make[3]: *** [Makefile:151: run_win32.obj] Error 2
make[3]: Leaving directory '/cygdrive/c/projects/ocaml/ocamltest'
```

Is this a known issue?

nojb commented:
`ocamltest` needs to be adapted to the Unicode changes introduced by #1200; I have a PR almost ready with the changes.  I am also investigating some kind of quoting issue (not sure what exactly yet) with `ocamltest` and `flexlink` under Windows that cause the tests to fail.

Will submit a PR as soon as it is ready.

shindere commented:
Hi,

Gabriel Scherer (2017/09/19 12:40 +0000):
> Currently, the AppVeyor CI machine seems broken (see for example [the fma build](https://ci.appveyor.com/project/avsm/ocaml/build/1.0.4243)), and I have the impression that this is ocamltest-related:
> 
> ```
> make[3]: Entering directory '/cygdrive/c/projects/ocaml/ocamltest'
> cl -nologo -O2 -Gy- -MD -WX -D_CRT_SECURE_NO_DEPRECATE -DCAML_NAME_SPACE -DUNICODE -D_UNICODE -DWINDOWS_UNICODE=1 -I../byterun -DCAML_INTERNALS  -c run_win32.c
> run_win32.c
> run_win32.c(61): error C2220: warning treated as error - no 'object' file generated
> run_win32.c(61): warning C4133: 'function': incompatible types - from 'const char *' to 'LPCWSTR'
> run_win32.c(64): warning C4133: 'function': incompatible types - from 'char *' to 'LPWSTR'
> run_win32.c(66): warning C4133: 'function': incompatible types - from 'char **' to 'LPWSTR *'
> run_win32.c(88): warning C4133: 'function': incompatible types - from 'const char *' to 'LPCWSTR'
> run_win32.c(91): warning C4133: 'function': incompatible types - from 'char *' to 'LPWSTR'
> run_win32.c(93): warning C4133: 'function': incompatible types - from 'char **' to 'LPWSTR *'
> run_win32.c(53): warning C4133: 'initializing': incompatible types - from 'char [5]' to 'LPCTSTR'
> run_win32.c(140): warning C4133: 'function': incompatible types - from 'const char *' to 'LPCWSTR'
> run_win32.c(157): warning C4133: 'function': incompatible types - from 'const char *' to 'LPCWSTR'
> run_win32.c(250): warning C4133: 'function': incompatible types - from 'char *' to 'LPCWSTR'
> run_win32.c(251): warning C4133: 'function': incompatible types - from 'char *' to 'LPWSTR'
> make[3]: *** [Makefile:151: run_win32.obj] Error 2
> make[3]: Leaving directory '/cygdrive/c/projects/ocaml/ocamltest'
> ```

Just to make sure. Does this problem still exist?

Thanks,

Sébastien.

gasche commented:
The problem doesn't show up anymore in the AppVeyor logs,
and I believe that it was fixed by nojb's #1357 (Fix ocamltest / Windows /
Unicode issue)

On Fri, Sep 29, 2017 at 11:15 AM, Sébastien Hinderer <
notifications@github.com> wrote:

> Hi,
>
> Gabriel Scherer (2017/09/19 12:40 +0000):
> > Currently, the AppVeyor CI machine seems broken (see for example [the
> fma build](https://ci.appveyor.com/project/avsm/ocaml/build/1.0.4243)),
> and I have the impression that this is ocamltest-related:
> >
> > ```
> > make[3]: Entering directory '/cygdrive/c/projects/ocaml/ocamltest'
> > cl -nologo -O2 -Gy- -MD -WX -D_CRT_SECURE_NO_DEPRECATE -DCAML_NAME_SPACE
> -DUNICODE -D_UNICODE -DWINDOWS_UNICODE=1 -I../byterun -DCAML_INTERNALS -c
> run_win32.c
> > run_win32.c
> > run_win32.c(61): error C2220: warning treated as error - no 'object'
> file generated
> > run_win32.c(61): warning C4133: 'function': incompatible types - from
> 'const char *' to 'LPCWSTR'
> > run_win32.c(64): warning C4133: 'function': incompatible types - from
> 'char *' to 'LPWSTR'
> > run_win32.c(66): warning C4133: 'function': incompatible types - from
> 'char **' to 'LPWSTR *'
> > run_win32.c(88): warning C4133: 'function': incompatible types - from
> 'const char *' to 'LPCWSTR'
> > run_win32.c(91): warning C4133: 'function': incompatible types - from
> 'char *' to 'LPWSTR'
> > run_win32.c(93): warning C4133: 'function': incompatible types - from
> 'char **' to 'LPWSTR *'
> > run_win32.c(53): warning C4133: 'initializing': incompatible types -
> from 'char [5]' to 'LPCTSTR'
> > run_win32.c(140): warning C4133: 'function': incompatible types - from
> 'const char *' to 'LPCWSTR'
> > run_win32.c(157): warning C4133: 'function': incompatible types - from
> 'const char *' to 'LPCWSTR'
> > run_win32.c(250): warning C4133: 'function': incompatible types - from
> 'char *' to 'LPCWSTR'
> > run_win32.c(251): warning C4133: 'function': incompatible types - from
> 'char *' to 'LPWSTR'
> > make[3]: *** [Makefile:151: run_win32.obj] Error 2
> > make[3]: Leaving directory '/cygdrive/c/projects/ocaml/ocamltest'
> > ```
>
> Just to make sure. Does this problem still exist?
>
> Thanks,
>
> Sébastien.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml/ocaml/pull/681#issuecomment-333075390>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAaA_mgpyr8TH07tW0-XFbRTyXOvhxt0ks5snLUtgaJpZM4JJbFr>
> .
>

shindere commented:
Okay thanks

objmagic commented:
@shindere seems `make one` is not working properly on ocamltests based test now . I am on `trunk` branch and `make one DIR=tests/typing-modules` does not run anything. 

This is reproducible on both my RedHat Linux machine with GNU Make 4.2.1 and my 10.12 OS X machine with GNU Make 3.8.1.

Seems the fix should be installed [here](https://github.com/ocaml/ocaml/blob/trunk/testsuite/Makefile#L203)

My attempted fix only works on my OS X machine and does not work on Linux. I don't know why and my Makefile writing skill is bad. I would appreciate if you can look into it.

objmagic commented:
`make promote` is not working either. We should also update `HACKING.adoc` under `testsuite`

v-gb commented:
Yep, neither is working with ocamltest. If you look at #1530, `make one` is not supposed to work with the new tests for some reason. #1519 also has some discussions about how `make promote` doesn't work, although it's actually not just expect-test, even tests-with-reference-files are broken in this way.
